kernel/src/hil/usb_hid.rs

```diff
@@ -0,0 +1,39 @@
+//! Interface for USB HID (Human Interface Device) class
+
+use crate::returncode::ReturnCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// Called when a packet is received
+    fn packet_received(&'a self, buffer: &'static mut [u8; 64], len: usize, endpoint: usize);
+
+    /// Called when a packet has been finished transmitting.
+    fn packet_transmitted(
+        &'a self,
+        result: Result<(), ReturnCode>,
+        buffer: &'static mut [u8; 64],
+        len: usize,
+        endpoint: usize,
+    );
+
+    /// Called when checking if we can receive any more data.
+    /// Should return true if we are ready to receive.
+    fn can_receive(&'a self) -> bool;
+}
+
+pub trait UsbHid<'a> {
```

Comment:
Unlike in the UART, I think it makes sense to have receive and transmit in the same trait here, because for USB the two are coupled, and for a particular endpoint we won't virtualize; virtualization would occur for higher layer abstractions, not at this level.

That being said, I think that these send/receive APIs should follow the same structure as the UART ones. The call to receive should pass the buffer. With this API, it's possible to lose a buffer, e.g., if there already is one and then `set_recv_buffer` is called again. Also, there are no error conditions returned. You need a way to cancel/halt receptions, rather than enabling them. So it should be `receive` and `cancel`, not `set_recv_buffer` and `allow_receive`. These APIs should clearly describe what the ownership assumptions are.

Why does `packet_transmitted` return the endpoint? It's not passed in `send_buffer`.

Reply:
I have converted this so that receive now takes a buffer (`receive_buffer()`) and there is a cancel call in the HIL (`receive_cancel()`).

I have added more documentation in the HIL as well.

`packet_transmitted()` returns the endpoint so that users know what endpoint the data was transfered over if there is more then one. It's not currently used, but it seems like it might be useful for more complex HID devices in the future.

Reply:
Looking at the implementation that @gendx did for OpenSK, it seems we even needed to be able to cancel both IN and OUT transactions for CTAP2.

Reply:
How would you cancel a send from Tock? If the transaction has already started you have already sent the length on the USB bus, you don't then want to cancel mid transaction.

The send command only sends a single buffer, I don't see a use case for when you would want to cancel that.

Reply:
As a USB peripheral you're not really deciding when you'll send the buffer. You can only enqueue it to be sent on the next transaction decided by the host, right? So conditions may occur when your enqueued buffer for transmission is not valid anymore because the host sent you another packet.
At least that would be my guess but Guillaume would know better than me here.

Reply:
That's true, but the USB implementation only has a single send buffer so in practice this gets send straight away.

I also don't really see a use case where the data would become invalid that quickly. This is a HID style device so the small payload is sent very quickly.

Reply:
In USB, nothing gets sent until the host decided to initiate an OUT transaction (for which we receive interrupts). One important scenario is to send a message fragmented into multiple packets to the host: for each packet, we have to wait for the host to initiate the transaction (first interrupt) and to complete it (second interrupt), before we move on to the next packet. Otherwise, we may end up writing the second packet while the first one is still in progress, i.e. we may end up overwriting a DMA buffer while it's still being read. And this can happen however small the send buffer and "straight away" data gets sent.

The issue is if the host doesn't initiate any OUT transaction. This can happen realistically in practice:
- The host sends a "cancel" message ahead of time rather than expecting a response. This happens when the user presses "cancel" in the Chrome browser popup when prompted for a touch on the security key for example.
- The host application crashed/terminated.

Without support for cancellation, we're stuck in these realistic scenarios:
- If we wait for interrupts, we can wait forever because no host application sends us an OUT transaction.
- If we don't wait for interrupts, and hope that it gets sent "straight away" we DO overwrite DMA buffers in practice.

I don't know the specifics of OpenTitan chip w.r.t. this scenarios, but given how the nRF52840 interrupts are set up, that's something we cannot ignore.

Another important scenario is to be able to handle both send and receive concurrently. Because as a Tock device we can expect to send a packet (OUT transaction), but actually the host decided to send us a packet instead - typically a "cancel" message in CTAP (IN transaction). Because USB is driven by the host, we cannot completely decouple sending and receiving, nor expect either sending or receiving to complete quickly (nor to ever complete).

Reply:
The buffers should not be overwritten as the data should not be changed until the `packet_transmitted()` callback is called, no matter how long it takes for the transfer.

So you are saying that if a user hits cancel during a CTAP operation from the host, that the data we want to send might never get polled by the host and we will be stuck? I think I'm missing how that can happen, the host still continuously polls the device for interrupt data. So the data should still be sent over USB, although it will then not be processed by the host.

I'll add a `send_cancel()` to the HIL so that in future if there is a reason to cancel a send it can be implemented.

Reply:
Marking as resolved as there is now a `send_cancel()` in the HIL.

Reply:
> I think I'm missing how that can happen, the host still continuously polls the device for interrupt data.

That's not what I've observed happening, at least on a Linux host. The host polls for interrupt data only when the userspace application requests it.